src/compiler/checker.ts

```diff
@@ -9311,11 +9311,8 @@ namespace ts {
                         getFlowTypeOfReference(accessExpression, propType) :
                         propType;
                 }
-                if (isTupleType(objectType)) {
-                    const restType = getRestTypeOfTupleType(objectType);
-                    if (restType && isNumericLiteralName(propName) && +propName >= 0) {
-                        return restType;
-                    }
+                if (everyType(objectType, isTupleType) && isNumericLiteralName(propName) && +propName >= 0) {
+                    return mapType(objectType, t => getRestTypeOfTupleType(<TupleTypeReference>t) || undefinedType);

```

Comment:
Hm, this isn't using a `undefinedWideningType` which means outside of `strict` mode, the new types are going to be strictly less permissive than they were before. Are we OK with that?

Reply:
They're going to be less permissive no matter which `undefined` we use because previously, when the object type was a union type, we'd just fall back to the numeric index signature. For example:

```ts
type T = [boolean] | [string, number];
type T0 = T[0];  // Unchanged, string | boolean
type T1 = T[1];  // Now number | undefined, previously number | boolean
type T2 = T[2];  // Now undefined, previously string | number | boolean
```

I would argue that the non-widening `undefined` is the right choice above. In non-strict mode, the only time you'd ever observe a difference is when it is not in a union with something else, i.e. `T2` above. I don't think we'd help anybody if `T2` was magically going to widen to `any`.